License text provider plugins

mnonnenmacher opened the pull request:
A draft for license text providers, created for early feedback. The plan is to convert the existing logic into separate providers that are queried in the order defined by the user, so two more will be added:
* ScanCode license text provider, to provider license texts from a local ScanCode installation.
* A provider for a directory containing custom license texts.
I would also like to add on that downloads license texts on demand from https://scancode-licensedb.aboutcode.org.

@sschuberth I just discovered #9620, so should this API be renamed to `LicenseDataProvider` already and the metadata model and functions could be added later?

TODO:

- [x] Take new API into use
- [x] Remove old API
- [x] Generate plugin docs
- [x] Write docs for license fact providers
- [x] Fix failing tests.

sschuberth commented:
> @sschuberth I just discovered #9620, so should this API be renamed to `LicenseDataProvider` already and the metadata model and functions could be added later?

If we agree that it should be the same interface / plugin type that returns more license (meta)data later (and not a different interface / plugin type), then I believe it would be good to signal that intention by already now generalizing the interface and class names to `LicenseDataProvider` or e.g. `LicenseMetadataProvider`.

As a side note, my old [`license-classification-interface` branch](https://github.com/oss-review-toolkit/ort/compare/main...sschuberth/license-classification-interface) contains a separate [`LicenseClassifications` interface](https://github.com/oss-review-toolkit/ort/compare/main...sschuberth/license-classification-interface#diff-c4ac9873caba7e0e7863a5c0180e2014dd9debf8f2ecca81e2111d4c3bee1412R24-R29) that could later be merged into the interface defined here (again, *if* we decide that texts and classifications should be provided by the same plugin type).

sschuberth commented:
Also interesting in this context: https://opensource.org/blog/introducing-the-new-api-for-osi-approved-licenses

mnonnenmacher commented:
The providers should now be on par with the currently used logic for reading license texts, but the integration is still missing. For ScanCode I have not copied the various fallbacks because I wasn't sure if they are still required.

sschuberth commented:
> For ScanCode I have not copied the various fallbacks because I wasn't sure if they are still required.

I believe we anyway should simplify the logic, and instead of looking in the various locations depending on how ScanCode was installed, we should use the `scancode-license-data` tool. A provider for it could call the tool once at runtime / initialization, and then serve the license texts from the location `scancode-license-data` was instructed to copy them to.

mnonnenmacher commented on plugins/license-fact-providers/spdx/src/main/kotlin/SpdxLicenseFactProvider.kt:
```diff
@@ -0,0 +1,41 @@
+
+    override fun hasLicenseText(licenseId: String) = getLicenseTextResource(licenseId) != null
+
+    private fun getLicenseTextResource(licenseId: String) = object {}.javaClass.getResource("/licenses/$licenseId")
```
I copied that from the existing function but I guess there it was only required because it was a top-level function.

sschuberth commented on plugins/license-fact-providers/api/src/main/kotlin/CompositeLicenseFactProvider.kt:
```diff
@@ -0,0 +1,45 @@
+    override fun getLicenseText(licenseId: String) =
+        providers.asSequence().mapNotNull { it.getLicenseText(licenseId) }.firstOrNull()
+
+    override fun hasLicenseText(licenseId: String) = providers.any { it.hasLicenseText(licenseId) }
```
Or we just do not introduce this complexity at all yet, but keep it in mind for a later (breaking) change...

mnonnenmacher commented:
@sschuberth Some tests are still failing but you could already make another review if you want.

sschuberth commented on plugins/reporters/asciidoc/src/main/resources/templates/freemarker_implicit.ftl:
Mind fixing this as well for `plugins/reporters/fossid/src/main/resources/templates/freemarker_implicit.ftl`?

sschuberth commented on plugins/license-fact-providers/spdx/build.gradle.kts:
```diff
@@ -0,0 +1,31 @@
+/*
```
Commit message: Typo in "repoter". Maybe also "enabled" -> "enables"?

sschuberth commented on plugins/license-fact-providers/spdx/src/test/kotlin/SpdxLicenseFactProviderTest.kt:
```diff
@@ -0,0 +1,78 @@
+        }
+
+        "return null for an unknown license" {
+            provider.getLicenseText("UnknownLicense") shouldBe null
```
Nit: Probably `should beNull()`.

sschuberth commented on plugins/license-fact-providers/scancode/src/funTest/kotlin/ScanCodeLicenseFactProviderFunTest.kt:
```diff
@@ -0,0 +1,85 @@
+class ScanCodeLicenseFactProviderFunTest : WordSpec({
+    "getLicenseText()" should {
+        "read license texts from the configured directory" {
+            val licenseDir = tempdir()
+            licenseDir.resolve("test.LICENSE").writeText("custom license text")
```
Nit: Maybe use `apply` (and / or the new `/` syntax).

sschuberth commented on plugins/license-fact-providers/scancode/src/funTest/kotlin/ScanCodeLicenseFactProviderFunTest.kt:
```diff
@@ -0,0 +1,85 @@
+        "read license texts from the detected ScanCode license directory" {
+            val provider = ScanCodeLicenseFactProviderFactory.create()
+
+            provider.getLicenseText("MIT").shouldNotBeNull {
```
Nit: Prefer infix `shouldNotBeNull` without the leading dot.

sschuberth commented on plugins/license-fact-providers/scancode/src/funTest/kotlin/ScanCodeLicenseFactProviderFunTest.kt:
```diff
@@ -0,0 +1,85 @@
+        "return null for an unknown license" {
+            val provider = ScanCodeLicenseFactProviderFactory.create(scanCodeLicenseTextDir = tempdir().absolutePath)
+
+            provider.getLicenseText("UnknownLicense") shouldBe null
```
`should beNull()`

sschuberth commented on plugins/license-fact-providers/scancode/src/main/kotlin/ScanCodeLicenseFactProvider.kt:
```diff
@@ -0,0 +1,131 @@
+     * The directory that contains the ScanCode license texts. If not set, the provider will try to locate the ScanCode
+     * license text directory using a heuristic based on the path of the ScanCode binary.
+     */
+    val scanCodeLicenseTextDir: String?
```
Probably not for now to keep this implementation closer to the current behavior, but would you agree that going forward we should probably change this implementation to lazy call `scancode-license-data` to expert ScanCode license data to a fixed / cache directory in favor of removing the heuristic?

sschuberth commented on plugins/license-fact-providers/scancode/src/main/kotlin/ScanCodeLicenseFactProvider.kt:
```diff
@@ -0,0 +1,131 @@
+     */
+    private val scanCodeLicenseTextDir: File by lazy {
+        if (config.scanCodeLicenseTextDir != null) {
+            File(config.scanCodeLicenseTextDir).takeIf { it.isDirectory }.also {
+                requireNotNull(it) {
+                    "Configured ScanCode license text directory '${config.scanCodeLicenseTextDir}' does not exist or " +
+                        "is not a directory."
+                }
+
+                logger.debug { "Using configured ScanCode license text directory: ${it.absolutePath}" }
+                return@lazy it
+            }
```
Maybe a little bit cleaner:

```kotlin
return@lazy File(config.scanCodeLicenseTextDir).also {
    require(it.isDirectory) {
        "Configured ScanCode license text directory '${config.scanCodeLicenseTextDir}' does not exist or " +
            "is not a directory."
    }

    logger.debug { "Using configured ScanCode license text directory: ${it.absolutePath}" }
}
```

sschuberth commented on plugins/license-fact-providers/scancode/src/main/kotlin/ScanCodeLicenseFactProvider.kt:
```diff
@@ -0,0 +1,131 @@
+        }
+
+        val licenseDir =
+            scanCodeBaseDir?.walkTopDown()?.find { it.isDirectory && it.endsWith("licensedcode/data/licenses") }
```
You could just use `walk()` as `FileWalkDirection.TOP_DOWN` is the default, which makes this line fit into one with the above.

sschuberth commented on plugins/license-fact-providers/scancode/src/main/kotlin/ScanCodeLicenseFactProvider.kt:
```diff
@@ -0,0 +1,131 @@
+            return@lazy licenseDir
+        }
+
+        File(FALLBACK_DIR).takeIf { it.isDirectory }.also {
```
How about making this variable a `File` instead of a `String` to begin with?

sschuberth commented on plugins/license-fact-providers/scancode/src/main/kotlin/ScanCodeLicenseFactProvider.kt:
```diff
@@ -0,0 +1,131 @@
+    }
+
+    private fun getLicenseTextFile(licenseId: String): File? {
+        val filename = "${licenseId.removePrefix("LicenseRef-").lowercase()}.LICENSE"
```
We should probably use the `LICENSE_REF_PREFIX` constant. Or didn't you want to introduce a dependency on `spdx-utils`?

sschuberth commented on plugins/license-fact-providers/dir/src/test/kotlin/DirLicenseFactProviderTest.kt:
```diff
@@ -0,0 +1,60 @@
+        "return null if no license text is found" {
+            val provider = DirLicenseFactProviderFactory.create(licenseTextDir = tempdir().absolutePath)
+
+            provider.getLicenseText("LicenseRef-non-existing-license") shouldBe null
```
`should beNull()`

sschuberth commented on utils/spdx/src/main/kotlin/Utils.kt:
```diff
@@ -105,68 +87,3 @@ fun calculatePackageVerificationCode(directory: File): String {
-            id.removePrefix(LICENSE_REF_PREFIX),
-            idWithoutLicenseRefNamespace,
-            "$idWithoutLicenseRefNamespace.LICENSE",
-            "x11-xconsortium_veillard.LICENSE".takeIf {
-                // Work around for https://github.com/aboutcode-org/scancode-toolkit/issues/2813 which affects ScanCode
-                // versions below 31.0.0.
-                id == "LicenseRef-scancode-x11-xconsortium-veillard"
-            }
```
I'm fine with dropping the workaround, but I'd appreciate this to be done as a preparing commit in the old code, so that it's more clear that this workaround was not accidentally dropped as part of the migration to the new API.

mnonnenmacher replied:
I've kept the workaround instead as we still officially support ScanCode >= 30.0.0.

mnonnenmacher commented:
@sschuberth In 8b19f2732f54b7357910ee41f7bd0af75ea9f7d8 I have changed the old license text provider to ignore the "+" in SPDX licenses. I have not done this in the new implementation which makes the `OpossumReporterFunTest` fail because:
* For some reason it includes ALL SPDX license texts in the output file.
* While the content of the license text resources with and without "+" seems to be identical, it is formatted differently (for example, compare [LGPL-3.0](https://github.com/oss-review-toolkit/ort/blob/main/utils/spdx/src/main/resources/licenses/LGPL-3.0) and [LGPL-3.0+](https://github.com/oss-review-toolkit/ort/blob/main/utils/spdx/src/main/resources/licenses/LGPL-3.0%2B)).

The reason for the commit was to fix #5004, however, in the meantime we have added license text resources for the licenses with "+" (I didn't check when). What's your opinion, should I update the expected results for the Opossum reporter or revert the SPDX provider to the previous behavior and ignore the "+"?

It's also weird that the formatting of the license texts for the "+" licenses is different to that of the according licenses without "+".

sschuberth commented on model/src/main/kotlin/config/OrtConfiguration.kt:
```diff
@@ -81,6 +83,17 @@ data class OrtConfiguration(
+     * license facts, license facts from a local ScanCode installation, and license facts from the default
+     * [ORT_CUSTOM_LICENSE_TEXTS_DIRNAME].
+     */
+    val licenseFactProviders: LinkedHashMap<String, PluginConfig> = linkedMapOf(
```
As `mapOf()` creates a `LinkedHashMap`, I still believe we should be fine with just `val licenseFactProviders: Map<String, PluginConfig> = mapOf(...)`.

mnonnenmacher replied:
On other hand it also makes sense to communicate in the Kotlin API that the order of the map is important, when changing the type to `Map` you could create the configuration programmatically with any kind of map potentially leading to unexpected results.

mnonnenmacher commented:
> > which makes the `OpossumReporterFunTest` fail because:
> > ```
> > * For some reason it includes ALL SPDX license texts in the output file.
> > ```
> 
> This has bothered me in the past in the context of other changes as well.
> 
> > * it is formatted differently (for example, compare [LGPL-3.0](https://github.com/oss-review-toolkit/ort/blob/main/utils/spdx/src/main/resources/licenses/LGPL-3.0) and [LGPL-3.0+](https://github.com/oss-review-toolkit/ort/blob/main/utils/spdx/src/main/resources/licenses/LGPL-3.0%2B)).
> 
> This formatting difference seems to come from the ScanCode license DB (compare [lgpl-3.0.html](https://scancode-licensedb.aboutcode.org/lgpl-3.0.html) to [lgpl-3.0-plus.html](https://scancode-licensedb.aboutcode.org/lgpl-3.0-plus.html)), not from the SPDX list (compare [LGPL-3.0-only.txt](https://github.com/spdx/license-list-data/blob/main/text/LGPL-3.0-only.txt) to [LGPL-3.0-or-later.txt](https://github.com/spdx/license-list-data/blob/main/text/LGPL-3.0-or-later.txt)), as historically the former had better formatting than the latter. Maybe it's time by now to really only use SPDX license texts in our resources. Or not use any resources / bundled license texts at all anymore, and least not in the core, but exclusively move this data to license fact provider plugins.
> 
> > What's your opinion, should I update the expected results for the Opossum reporter or revert the SPDX provider to the previous behavior and ignore the "+"?
> 
> I'd not ignore the "+" anymore; after all, the license texts for these variants _could_ be different. But I believe we should always fall back to the non-"+" text if there is no dedicated "+" text. The question is whether e should leave this for each provider to implement, or guarantee this fallback somehow via the API.
> 
> So regarding `OpossumReporterFunTest`, I'd either update the expected results, or even reduce the test case to at least not test any "+" variants, or maybe even trim it down more to only check a few selected license texts.

I have updated the expected results because that was the easiest solution.

mnonnenmacher commented on model/src/main/kotlin/config/OrtConfiguration.kt:
```diff
@@ -81,6 +83,17 @@ data class OrtConfiguration(
+     * [ORT_CUSTOM_LICENSE_TEXTS_DIRNAME].
+     */
+    val licenseFactProviders: LinkedHashMap<String, PluginConfig> = linkedMapOf(
+        "SPDX" to PluginConfig.EMPTY,
+        "ScanCode" to PluginConfig.EMPTY,
```
So is the argument that the formatting of the ScanCode license texts is better not valid anymore?
